File: A09 two pointer(sliding Window)/TwoSum or KSum/LeetCode 18 4Sum/two pointer.py
###O(N^3) time complixity 
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        def find_quadruples(nums, first, second, quadruples):
            left = second+1
            right = len(nums)-1
            while left < right:
                sum = nums[first]+nums[second]+nums[left]+nums[right]
                if sum == target:
                    quadruples.append([nums[first], nums[second], nums[left], nums[right]])
                    left += 1
                    right -= 1
                    while left < right and nums[left] == nums[left-1]:
                        left += 1
                    while left < right and nums[right] == nums[right+1]:
                        right -= 1
                elif sum < target:
                    left += 1
                else: right -= 1
        
        if len(nums) < 4: return []
        quadruples = []
        nums.sort()
        for i in range(len(nums)-3):
            if i > 0 and nums[i] == nums[i-1]: continue
            for j in range(i+1,len(nums)-2):
                if j > i+1 and nums[j] == nums[j-1]: continue
                find_quadruples(nums, i, j, quadruples)
        return quadruples

File: A09 two pointer(sliding Window)/TwoSum or KSum/LeetCode 18 4Sum/two pointer.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        //corner case 
        if (nums == null || nums.length < 4) return new ArrayList<>();
        
        List<List<Integer>> res = new ArrayList<>();
        //int target = 0;
        Arrays.sort(nums);
        
        for(int i = 0 ; i < nums.length - 3; i ++){
            if (i > 0 && nums[i] == nums[i -1]) continue;    //skip if current and previous are same number
            for (int j = i + 1 ; j < nums.length -2; j++){
                if (j > i +1 && nums[j] == nums[j -1]) continue;  //skip duplicte
                int k = j +1; 
                int l = nums.length - 1; 
            
                while(k < l){
                    if (nums[i] + nums[j] + nums[k] + nums[l] < target){
                        k ++; 
                        while (k < l && nums[k] == nums[k -1]) k++;    //skip duplicates
                    }else if(nums[i] + nums[j] + nums[k] + nums[l] > target){
                        l --; 
                        while (k < l && nums[l] == nums[l+1]) l--;      //skip duplicates
                    
                    }else{
                        res.add(Arrays.asList(nums[i], nums[j], nums[k], nums[l])); 
                    
                        k++;                               //continue rest since find all sets
                        l--;
                        while (k < l && nums[k] == nums[k -1]) k++; 
                        while (k < l && nums[l] == nums[l+1]) l--; 
                    }
                }
            }
        }
        return res; 
    }
}



##################